### ISSUE TITLE
Identify lncRNAs in Geoduck Genome

### ISSUE THREAD

**Issue by sr320** [RobertsLab/resources#1434]

**Comment by sr320** [comment]
Using RNA-seq files used to generate https://gannet.fish.washington.edu/Atumefaciens/20220323-pgen-nextflow_rnaseq-tissues/star_salmon/salmon.merged.gene_tpm.tsv

take through general workflow https://zbengt.github.io/2023-04-20-LncRNA-Discovery/

with idea there would be 5 resulting lncRNA tracks that would be specific to the 5 "tissues" headers of star_salmon table above...

**Comment by kubu4** [comment]
@zbengt - Where can I find your input/output files for your lncRNA workflow that Steven linked above?

**Comment by zbengt** [comment]
For that workflow everything is here /home/shared/8TB_HDD_02/zbengt/github/zach-lncRNA in either the data or output directory. But in scaling up I put all of the paired reads from [here](https://gannet.fish.washington.edu/Atumefaciens/hputnam-Becker_E5/Becker_RNASeq/data/raw/) on Raven at this location: /home/shared/8TB_HDD_01/pver. I saw the issue from earlier today, I have yet to load any data onto mox.

**Comment by sr320** [comment]
@zbengt can you also provide the URL to your Rmarkdown code?

**Comment by zbengt** [comment]
Yep here is the [Rmd code](https://github.com/course-fish546-2023/zach-lncRNA/blob/main/code/03-get-fasta.Rmd). And here is my [lab notebook post](https://zbengt.github.io/2023-04-20-LncRNA-Discovery/). Note that I used the web version of [CPC2](http://cpc2.gao-lab.org/).

**Comment by kubu4** [comment]
@zbengt - Link to R markdown doesn't work.

**Comment by sr320** [comment]
@zbengt can you make your repo public

**Comment by zbengt** [comment]
Sorry, repo was private, try again [here](https://github.com/course-fish546-2023/zach-lncRNA/blob/main/code/03-get-fasta.Rmd)

**Comment by kubu4** [comment]
Quick update:

Didn't identify any lncRNAs in geoduck data (i.e. annotated GTF from gffcompare has no entries annotated with `class_code "u"`).

Will review steps and report back.

**Comment by kubu4** [comment]
Gah! Think I figured it out. Had a leftover option enabled (`-e`) from some recycled StringTie code which limits output to only transcripts overlapping reference transcripts.

Re-running now. Should be finished in ~6hrs.

**Comment by kubu4** [comment]
@zbengt -

What does this line do:

`awk '$NF < 0 {print $1}' ../output/cpc2results-last.txt`

Specifically, what is the `$NF`? Usually, `awk` uses the `NF` variable (notice _no_ dollar sign) to count the number of fields in a line. So, I'm not sure what the dollar sign represents here?

Also, the final column of `cpc2results-last.txt` categorizes sequences as `coding/noncoding`. Would it be sufficient to utilize this column for parsing?

Finally, your notebook entry says to parse out entries with a "score less than zero". Which score? The `Fickett_score` or the `coding_probability`?

**Comment by zbengt** [comment]
TL;DR Info dump from our slack conversation:

This line prints the first column of each line where the last column's value is less than zero. Basically capturing transcript IDs based on coding potential score (1 or -1).
 
```
awk '$NF < 0 {print $1}' ../output/cpc2results-last.txt
```
I believe $NF indicates that you want the information within each cell of that column, rather than the class or count of that column. So in this case, in captures the transcript ID.

I was using what I though was the primary proxy for coding potential (ORF integrity, column 6), which is -1 for noncoding and 1 for coding. Parsing by "scores less than 0" in that column just chooses those with a score of -1 (noncoding). But definitely need to revisit this, likely using the final column `label` that classifies as coding or noncoding.

We've been using [this](https://link.springer.com/content/pdf/10.1007/978-1-4939-9045-0_13.pdf) paper for reference. But there are differences since they used the web-based version of CPC rather than the CPC2 like we are using, leading to some of our confusion.

And [here](https://github.com/course-fish546-2023/zach-lncRNA/blob/main/code/07-scale-up.Rmd) is my most recent, scaled-up code so far.

**Comment by kubu4** [comment]
To clarify some of the confusion with my questions to @zbengt:

I made an incorrect assumption about his CPC2 output file (`cpc2results-last.txt`), which was based off of _my_ CPC2 output file. 

It turns out, his CPC2 output file, generated using the web interface, looks like this:

|    |  |  |  |  |  |   |
|------------------------------------------------------|-----|---------|-------------|-----------|-----------|----|
| transcript::Pver_Sc0000000_size2095917:126924-127873 | 121 | 0.34477 | 9.984802246 | 0.291075  | noncoding | 1  |
| transcript::Pver_Sc0000000_size2095917:205923-206376 | 32  | 0.3095  | 7.835510254 | 0.100248  | noncoding | -1 |
| transcript::Pver_Sc0000000_size2095917:405496-413565 | 107 | 0.28467 | 11.31854248 | 0.0918163 | noncoding | 1  |

Whereas mine, from the CPC2 standalone/command line, looks like this:

| #ID                                   | transcript_length | peptide_length | Fickett_score | pI                 | ORF_integrity | coding_probability | label     |
|---------------------------------------|-------------------|----------------|---------------|--------------------|---------------|--------------------|-----------|
| MSTRG.38.5::Scaffold_01:656906-657583 | 677               | 72             | 0.30393       | 8.63104343414307   | 1             | 0.0370878          | noncoding |
| MSTRG.39.1::Scaffold_01:648204-649326 | 1122              | 48             | 0.32127       | 10.358344841003415 | 1             | 0.0232355          | noncoding |
| MSTRG.63.1::Scaffold_01:849165-854552 | 5387              | 64             | 0.30411       | 7.6276449203491214 | 1             | 0.0265552          | noncoding |


When reviewing his code, I neglected to look at his actual file and was basing my questions based off of my file - assuming we had the same file layout. This was wrong.

The output from the command line version does not have a final column with a coding probability score (`1/-1`).

Sorry about this!

**Comment by sr320** [comment]
@kubu4 how close did chatGPT get?

This code is a Jupyter Notebook designed to identify long non-coding RNAs (lncRNAs) from a given annotated GTF file. The GTF file is generated by gffcompare from the Panopea-generosa-v1.0 genome annotation. The analysis uses CPC2 Standalone to identify lncRNAs with no coding potential. A minimum length cutoff of 200 base pairs is applied for lncRNAs.

The code proceeds with the following steps:

List computer specs.
Set variables for input files, output files, and program locations.
Create an analysis directory.
Download the GTF file and inspect it.
Download the genome FastA file and inspect it.
Parse out transcripts that are equal to or greater than 200bp and have no overlap with reference transcripts.
Convert the GTF file to BED format.
Extract FastA sequences from candidate lncRNAs using bedtools.
Run CPC2 to identify non-coding RNAs.
Capture noncoding IDs from the CPC2 output table.
Extract lncRNAs from the GTF file.
Generate checksums for output files.
Document GffRead program options.
The final output is a GTF file containing lncRNAs with no coding potential and a minimum length of 200 base pairs.

**Comment by kubu4** [comment]
Ha! Nice summary!

Looks like I did a great job of annotating the notebook so that it could generate this. :)

**Comment by kubu4** [comment]
@zbengt - Check out [this comment](https://github.com/RobertsLab/resources/issues/1642#issuecomment-1535333952) for lncRNA expression.

**Comment by zbengt** [comment]
Very cool! Thanks Sam!
